flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperator.java

```diff
@@ -319,11 +319,14 @@ public void merge(W mergeResult,
 									"event-time window cannot become earlier than the current watermark " +
 									"by merging. Current watermark: " + internalTimerService.currentWatermark() +
 									" window: " + mergeResult);
-						} else if (!windowAssigner.isEventTime() && mergeResult.maxTimestamp() <= internalTimerService.currentProcessingTime()) {
-							throw new UnsupportedOperationException("The end timestamp of a " +
+						} else {
```

Comment:
This introduces an additional call to `internalTimerService.currentProcessingTime()`, which does a syscall, if I'm not mistaken, for the current time every time the else branch is taken.

We should probably do the `!windowAssigner.isEventTime()` check in the else and then retrieve the current processing time within the else block.